Update Rust module documentation (#2050)
clockworklabs/SpacetimeDB: crates/bindings/README.md

> # SpacetimeDB Rust Module Library

Author: I recommend reviewing starting here

┌───────────────────────┐                ┌───────────────────────┐
│                       │                │                       │
│  ┌─────────────────┐  │    SQL Query   │  ┌─────────────────┐  │
│  │ Subscribed Data │<─────────────────────│    Database     │  │
│  └─────────────────┘  │                │  └─────────────────┘  │
│           │           │                │           ^           │
│           │           │                │           │           │
│           v           │                │           v           │
│  +─────────────────┐  │ call_reducer() │  ┌─────────────────┐  │
│  │   Client Code   │─────────────────────>│   Module Code   │  │
│  └─────────────────┘  │                │  └─────────────────┘  │
│                       │                │                       │
└───────────────────────┘                └───────────────────────┘
```

Rust modules are written with the the Rust Module Library (this crate). They are built using [cargo](https://doc.rust-lang.org/cargo/) and deployed using the [`spacetime` CLI tool](https://spacetimedb.com/install). Rust modules can import any Rust [crate](https://crates.io/) that supports being compiled to WebAssembly.

> (Note: Rust can also be used to write **clients** of SpacetimeDB databases, but this requires using a completely different library, the SpacetimeDB Rust Client SDK. See the documentation on [clients] for more information.)

Author: There's an anchor at the bottom of the file.

> (Note: Rust can also be used to write **clients** of SpacetimeDB databases, but this requires using a completely different library, the SpacetimeDB Rust Client SDK. See the documentation on [clients] for more information.)

Reviewer: D'oh! Not sure how I missed that.

  - (Or creates a fresh database and identity, if no identity was provided).
- The host loads the new module and inspects its requested database schema. If there are changes to the schema, the host tries perform an [automatic migration](#automatic-migrations). If the migration fails, publishing fails.
- The host terminates the old module attached to the database.
- The host installs the new module into the database. It begins running the module's [lifecycle reducers](#lifecycle-reducers) and [scheduled reducers](#scheduled-reducers).
- The host begins allowing clients to call the module's reducers.

From the perspective of clients, this process is mostly seamless. Open connections are maintained and subscriptions continue functioning. [Automatic migrations](#automatic-migrations) forbid most table changes except for adding new tables, so client code does not need to be recompiled.
However:
- Clients may witness a brief interruption in the execution of scheduled reducers (for example, game loops.)
- New versions of a module may remove or change reducers that were previously present. Client code calling those reducers will receive runtime errors.

## Tables

> Tables are declared using the [`#[table(name = table_name)]` macro](macro@crate::table).

Author: It is a rustdoc link: https://rust-lang.github.io/rfcs/1946-intra-rustdoc-links.html#path-ambiguities

This macro is applied to a Rust struct with named fields. All of the fields of the table must implement [`SpacetimeType`].

> The resulting type is used to store rows of the table. It is normal struct type. Row values are not special -- operations on row types do not, by themselves, modify the table. Instead, a [`ReducerContext`](#reducercontext) is needed to get a handle to the table.

Author: I think it's important to clarify exactly what magic is happening.

> The resulting type is used to store rows of the table. It is normal struct type. Row values are not special -- operations on row types do not, by themselves, modify the table. Instead, a [`ReducerContext`](#reducercontext) is needed to get a handle to the table.

Author: Additionally, I'd propose https://github.com/clockworklabs/SpacetimeDB/issues/2261

Any `#[unique]` or `#[primary_key]` column supports getting a [`UniqueColumn`] from a [`ReducerContext`] using:

```text
ctx.db.{table}().{unique_column}()
```

For example, 

```no_run
ctx.db.people().ssn()
```
    
[`UniqueColumn`] provides:
- [`UniqueColumn::find`]
- [`UniqueColumn::delete`]
- [`UniqueColumn::update`]
<!-- TODO: "current limitations" try_update -->

Notice that updating a row is only possible if a row has a unique column -- there is no `update` method in the base [`Table`] trait. SpacetimeDB has no notion of rows having an "identity" aside from their unique / primary keys.

The `#[primary_key]` annotation is similar to the `#[unique]` annotation, except that it leads to additional methods being made available in the [client]-side SDKs. <!-- TODO: better link. -->

<!-- TODO: does this actually work? Are there still limitations on what we can mark unique?

It is not currently possible to mark a group of fields as collectively unique. You can, however, derive `#[SpacetimeType]` on a sub-struct and mark that unique. For example:

...

> -->

Reviewer: Oh, it seems like we actually *don't* impose this limit on unique columns in Rust, only non-unique BTree indexes. That's... not ideal.

> -->

Reviewer: Fixed in #2267 . I will update the docs here.

```rust
use spacetimedb::{reducer, ReducerContext};

#[derive(Debug)]
enum GiveItemError {
    NoSuchItem(u64),
    NonTransferable(u64)
}

#[reducer]
fn give_player_item(
    ctx: &ReducerContext,
    player_id: u64,
    item_id: u64
) -> Result<(), GiveItemError> {
    /* ... */
}
```

Every reducer runs inside a [database transaction](https://en.wikipedia.org/wiki/Database_transaction). <!-- TODO: specific transaction level guarantees. --> This means that reducers will not observe the effects of other reducers modifying the database while they run. Also, if a reducer fails, all of its changes to the database will automatically be rolled back. Reducers can fail by [panicking](::std::panic!) or by returning an `Err`.

#### The `ReducerContext` Type

Reducers have access to a special [`ReducerContext`] argument. This argument allows reading and writing the database attached to a module. It also provides some additional functionality, like generating random numbers and scheduling future operations.

> The most important field of [`ReducerContext`] is [`.db`](ReducerContext#structfield.db). This field provides a view of the module's database. The [`#[table]`](macro@crate::table) macro generates traits that add accessor methods to this field.

Reviewer: "The most important field" seems like a silly thing to say, IMO. Does `ctx.sender` not matter? Should `ctx.identity()` just go screw itself?

```suggestion
[`ReducerContext`] provides access to the database tables via [the `.db` field](ReducerContext#structfield.db). The [`#[table]`](macro@crate::table) macro generates traits that add accessor methods to this field.
```

Author: ok but

```
james@DESKTOP-1O4DK4H MINGW64 /f/clockwork/SpacetimeDB/modules (jgilles/rust-docs-update)
$ rg 'ctx.db' --stats | grep matches
159 matches
8 files contained matches

james@DESKTOP-1O4DK4H MINGW64 /f/clockwork/SpacetimeDB/modules (jgilles/rust-docs-update)
$ rg 'ctx.sender' --stats | grep matches
14 matches
4 files contained matches

james@DESKTOP-1O4DK4H MINGW64 /f/clockwork/SpacetimeDB/modules (jgilles/rust-docs-update)
$ rg 'ctx.identity' --stats | grep matches
2 matches
1 files contained matches
```

:3

<!-- TODO: this seems to work sometimes, but not always... Sometimes the links to downstream trait implementations aren't generated for some reason. Maybe it only works in the same cargo workspace?

To see all of the available methods on `ctx.db`, run `cargo doc` in your module's directory, and navigate to the `spacetimedb::Local` struct in the generated documentation. This will be at the path:
- `[your_project_directory]/target/doc/spacetimedb/struct.Local.html` (non-Windows)

> - `[your_project_directory]\target\doc\spacetimedb\struct.Local.html` (Windows)

Reviewer: I don't think you actually need to distinguish Windows vs Unix paths in 2025 Anno Domini. Just use `/` as a path separator.

Author: old habits

-->

#### The `log` crate

SpacetimeDB Rust modules have built-in support for the [log crate](log). All modules automatically install a suitable logger when they are first loaded by SpacetimeDB. (At time of writing, this happens [here](https://github.com/clockworklabs/SpacetimeDB/blob/e9e287b8aab638ba6e8bf9c5d41d632db041029c/crates/bindings/src/logger.rs)). Log macros can be used anywhere in module code, and log outputs of a running module can be inspected using the `spacetime logs` command:

```text
spacetime logs <DATABASE_IDENTITY>
```

<!-- TODO: we seem to have removed `println`, is that still hidden somewhere? -->

#### Lifecycle Reducers

A small group of reducers are called at set points in the module lifecycle. These are used to initialize
the database and respond to client connections. See [Lifecycle Reducers](macro@crate::reducer#lifecycle-reducers).

#### Scheduled Reducers

Reducers can be scheduled to run repeatedly. This can be used to implement timers, game loops, and

> maintenance tasks. See [Scheduled Reducers](macro@crate::reducer#scheduled-reducers).

Reviewer: Is scheduled-ness a feature of the reducer, or of the table? If the latter, we may want to document this as "scheduled tables" or something. I'm unsure on what's best.

Reviewer: I'm deciding I don't really care. This is fine.

SpacetimeDB attempts to automatically migrate your existing database to the new schema. (The "schema" is just the collection
of tables and reducers you've declared in your code, together with the types they depend on.) This form of migration is very limited and only supports a few kinds of changes.
On the plus side, automatic migrations usually don't break clients. The situations that may break clients are documented below.

The following changes are always allowed and never breaking:

<!-- TODO: everything here should be smoke-tested. -->

- ✅ **Adding tables**. Non-updated clients will not be able to see the new tables.
- ✅ **Adding indexes**.
- ✅ **Adding or removing `#[autoinc]` annotations.**
- ✅ **Changing tables from private to public**.
- ✅ **Adding reducers**.
- ✅ **Removing `#[unique]`  annotations.**

The following changes are allowed, but may break clients:

- ⚠️ **Changing or removing reducers**. Clients that attempt to call the old version of a changed reducer will receive runtime errors.
- ⚠️ **Changing tables from public to private**. Clients that are subscribed to a newly-private table will receive runtime errors.

> - ⚠️ **Removing `#[primary_key]` annotations**. Non-updated clients will still use the old `#[primary_key]` as a unique key in their local cache, which can result in non-deterministic behavior when updates are received.

Reviewer: I don't think this breaks clients. It's certainly not non-deterministic. It would be a breaking change to downgrade `primary_key` to not-unique-at-all, but I think we only allow `primary_key` -> `unique`, which doesn't actually change client behavior at all.

Author: Removing `#[unique]` is allowed, so I don't see any logic that would catch what you're saying. 

...And the `primary_key` field is ignored by `auto_migrate_table`, whoops! https://github.com/clockworklabs/SpacetimeDB/issues/2262

> - ⚠️ **Removing indexes**. This is only breaking in some situtations.

Reviewer: Do we support removing unique indexes (which come with unique constraints)?

Author: Yes.

  The specific problem is subscription queries <!-- TODO: clientside link --> involving semijoins, such as:

>     ```text

Reviewer: ```suggestion
    ```sql
```

    SELECT *
    FROM Employee
    WHERE DeptName IN (
    SELECT DeptName
    FROM Dept

>     )

Reviewer: I don't know if we actually support this construction of semijoins. (CC @joshua-spacetime ). We've been using the equivalent:
```sql
SELECT Employee.* FROM Employee JOIN Dept ON Employee.DeptName = Dept.DeptName
```
We also only support semijoins that "look like" PK/FK joins, i.e. where both columns are unique (* I don't actually know if we enforce this, or just assume it. @joshua-spacetime may know more), and where there's an additional `WHERE` filter on the "filter table" (`Dept` in this case) (* This should be changing in the future, but I believe this is the current status).

Reviewer 2: Phoebe is correct, we don't support `IN`, so
```sql
SELECT Employee.* FROM Employee JOIN Dept ON Employee.DeptName = Dept.DeptName
```
is the correct way to express it atm.

As of 1.0, a `WHERE` clause is no longer required.

Regarding `PK/FK` joins, previously the server did not enforce join columns having unique constraints, but the client could produce incorrect results in such cases. However, with the new 1.0 subscription api, the client does/will have handling for non-unique join columns.
